src/hotspot/share/gc/g1/g1Policy.cpp

```diff
@@ -781,7 +785,10 @@ void G1Policy::record_collection_pause_end(double pause_time_ms) {
   assert(!(is_concurrent_start_pause(this_pause) && collector_state()->mark_or_rebuild_in_progress()),
          "If the last pause has been concurrent start, we should not have been in the marking window");
   if (is_concurrent_start_pause(this_pause)) {
-    collector_state()->set_mark_or_rebuild_in_progress(true);
+    if (!concurrent_operation_is_full_mark) {
+      abort_time_to_mixed_tracking();
+    }
```

Comment:
More or less same idea here. What do you think about moving this to a separate function, that is either called from `start_concurrent_cycle()` or `concurrent_undo_cycle_do()`?

Reply:
The reason for this abort_time_to_mixed_tracking() call is that we need to undo what we've done in record_pause() earlier in the same method. That is why the change does the undo of that part here.

Also, this would place policy state changes outside of the application of the gc algorithm to the heap (I consider changes that are between the gc_prologue() and gc_epilogue() "young collection algorithm"). Putting this would break that assumption.

I fixed this in the most recent change by making the concurrent start pause that starts an undo explicit so that record_pause() does not start .